core/commonMain/src/kotlinx/serialization/internal/ElementMarker.kt

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright 2017-2021 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package kotlinx.serialization.internal
+
+import kotlinx.serialization.ExperimentalSerializationApi
+import kotlinx.serialization.descriptors.SerialDescriptor
+import kotlinx.serialization.encoding.CompositeDecoder
+
+@OptIn(ExperimentalSerializationApi::class)
+@PublishedApi
+internal class ElementMarker(
+    private val descriptor: SerialDescriptor,
+    private val filter: (SerialDescriptor, Int) -> Boolean
```

Comment:
`filter` name is too broad imo. Better to use something like `shouldReturn(Mark)Index(Element)` or just reuse `readElementIfAbsence`